DIAN-129

dephillipsmichael opened the pull request:
It's important to read the **ReadMe.md** changes in this pull request first before reviewing the code for full context about the migration code algorithm.

Unit test coverage below, with most of the missed methods being ones that directly call the Synapse API or Bridge API.
<img width="830" alt="Screen Shot 2021-10-17 at 10 54 23 AM" src="https://user-images.githubusercontent.com/5590748/137639094-fbe15ec6-ddee-4154-bce6-c2efdfec1686.png">

As for manual testing... after creating all the relevant sub-studies on the **Arc Validation** bridge project,
I have run the migration code pointed at this [Synapse project](https://www.synapse.org/#!Synapse:syn25791302/files/) a few times.  The first run successfully created all the relevant users, and then before the second run, I marked a few users as migrated, and the migration code properly deleted the user's temporary data.

dephillipsmichael committed:
Update gitignore for android studio

dephillipsmichael committed:
Initial commit with working CI integration tests and main executable

dephillipsmichael committed:
Remove action on ci_test branch

dephillipsmichael committed:
Inline secret

dephillipsmichael committed:
Bridge Rest API and additional integration test coverage

dephillipsmichael committed:
Updated CI

dephillipsmichael committed:
Added integration test

dephillipsmichael commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,573 @@
+    // Do not change the name of these two variables, as it is used in CI to replace the values
+    // However, to get the app to run, you must provide your own email/password,
+    // NEVER commit your credentials to github
+    public static String bridgeAdminEmail = "";
```
@brucehoff I also had a question about security. The JAR file will need to store both the Synapse personal token and bridge admin email/password, and it currently does that by using github secrets to inject the values into the java files like here. If someone were to get ahold of the JAR, they could de-compile and read those secrets.  Is there a more secure way you know of doing this?  Does docker compile everything and create the JAR so that it doesn't live outside the scope of the secure sage server?

brucehoff replied:
@dephillipsmichael The .jar file (like any source code or compiled code artifact) absolutely should NOT have any secrets stored in it.  A good way to pass a secret to a Java app at run time is via an "environment variable".  If the artifact is a Docker container, environment variables can be passed into the container using the "-e" option.  (See https://www.cloudsavvyit.com/14081/how-to-pass-environment-variables-to-docker-containers/) GitHub actions lets you pass secrets as environment variables at run time.

dephillipsmichael committed:
Added new UserMigration and UserQA scripts to control migrating user accounts

brucehoff commented on .github/workflows/unit_tests_and_jar.yml:
```diff
@@ -0,0 +1,41 @@
+name: Unit Tests
```
If this is a set of "unit tests", why are access tokens retrieved?  I would have thought that access tokens are only needed for integration tests.

dephillipsmichael replied:
Mislabeled on my part, they were integration tests.  I'm going to do a test refactor to switch over to actual unit tests to remove any dependency on a server.

brucehoff commented on .github/workflows/unit_tests_and_jar.yml:
```diff
@@ -0,0 +1,41 @@
+name: Unit Tests
+
+on:
+  pull_request:
```
Running a job on a pull request which retrieves access tokens is a security hole:  Someone can send a PR containing code that simply exfiltrates the secrets.  It's OK to run unit tests on a PR, but integration tests using access tokens should only be run on code reviewed by a person (as well as on your own fork before creating the PR).

dephillipsmichael replied:
Good insight.  I'm switching over to env vars as suggested.

brucehoff commented on .github/workflows/unit_tests_and_jar.yml:
```diff
@@ -0,0 +1,41 @@
+        
+      - name: Synapse access token copy
+        run: |
+          sed -i 's/synapsePersonalAccessToken = ""/synapsePersonalAccessToken = "${{ secrets.SYNAPSE_PAT }}"/' ${GITHUB_WORKSPACE}/DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java
```
No, please pass an an env variable.  You can retrieve it using System.getenv(name); where name is something like "synapseAccessToken".

brucehoff commented on .github/workflows/unit_tests_and_jar.yml:
```diff
@@ -0,0 +1,41 @@
+          
+      - name: Bridge auth copy 1
+        run: |
+          sed -i 's/bridgeAdminEmail = ""/bridgeAdminEmail = "${{ secrets.BRIDGE_EMAIL }}"/' ${GITHUB_WORKSPACE}/DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java
```
No, please pass an an env variable.

brucehoff commented on .github/workflows/unit_tests_and_jar.yml:
```diff
@@ -0,0 +1,41 @@
+          
+      - name: Bridge auth copy 2
+        run: |
+          sed -i 's/bridgeAdminPassword = ""/bridgeAdminPassword = "${{ secrets.BRIDGE_PW }}"/' ${GITHUB_WORKSPACE}/DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java
```
No, please pass an an env variable.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+import java.util.List;
+import java.util.Map;
+
+public class BridgeUtil {
```
FWIW we have Java client library, you don't need to work with JSON, HTTP, etc. if you don't want.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+        return responseObj.sessionToken;
+    }
+
+    public static <T> T getSingletonReport(
```
The Participant Data API does not create time series data, you can save a report by an ID for any user:

https://developer.sagebridge.org/swagger-ui/index.html#/Participant%20Data

What you're doing will work, but it's unnecessary.

dephillipsmichael replied:
I didn't know about this API, is it new?  

The Sage mobile team has been using this paradigm for years to store "singleton" reports which is where I got the idea.  Hopefully this will be used in the new mobile sdk.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+     */
+    public static List<MigrationPair> getUsersToMatch(
+            List<SynapseUtil.HmUserData> hmUsers,
+            List<BridgeUserData> bridgeUserList) {
```
Seems like this is more useful as a map from some identifier to the user record

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/UnzipUtil.java:
```diff
@@ -0,0 +1,96 @@
+            zipIn.closeEntry();
+            entry = zipIn.getNextEntry();
+        }
+        zipIn.close();
```
put close() in a 'finally' block, or use  a [try-with-resource](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html), which does that implicitly.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/UnzipUtil.java:
```diff
@@ -0,0 +1,96 @@
+        while ((read = zipIn.read(bytesIn)) != -1) {
+            bos.write(bytesIn, 0, read);
+        }
+        bos.close();
```
put close() in a 'finally' block, or use  a [try-with-resource](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html), which does that implicitly.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+    // Do not change the name of this variable, as it is used in CI to replace the token
+    // However, to get the app to run, you must provide your own synapse PAT,
+    // NEVER commit your personal access token to github
+    public static String synapsePersonalAccessToken = "";
```
pass token as an environment variable

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+    public static SynapseClient synapse;
+
+    private static Project project;
+    private static String projectId = "syn25791302";
```
pass projectId as an env variable to allow us to switch from UAT to production.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                synapse.downloadFile(file, download);
+                System.out.println("Unzipping file " + file.getAssociateObjectId() + ".zip");
+                UnzipUtil.unzip(download.getAbsolutePath(),
+                        testSessionScheduleExtractedFolder.getAbsolutePath());
```
I see that multipole .zip files are unzipped into the same folder.  Is there a chance of overwriting, i.e. can two zips have the same file name?

dephillipsmichael replied:
Happy Medium puts the date as a suffix in all filenames, so we can rely on all filenames being unique.

dephillipsmichael replied:
I'll add a comment in the code to make that more clear.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                synapse.downloadFile(file, download);
+                System.out.println("Unzipping file " + file.getAssociateObjectId() + ".zip");
+                UnzipUtil.unzip(download.getAbsolutePath(),
+                        testSessionExtractedFolder.getAbsolutePath());
```
I see that multipole .zip files are unzipped into the same folder. Is there a chance of overwriting, i.e. can two zips have the same file name?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                System.out.println("Downloading file " + file.getAssociateObjectId() + ".zip");
+                synapse.downloadFile(file, download);
+                System.out.println("Unzipping file " + file.getAssociateObjectId() + ".zip");
+                UnzipUtil.unzip(download.getAbsolutePath(),
```
I see that multipole .zip files are unzipped into the same folder. Is there a chance of overwriting, i.e. can two zips have the same file name?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+        File hasdParticipantsFolder = createFolderIfNecessary(HASD_PARTICIPANTS_FOLDER_NAME);
+        File hasdParticipantsExtractedFolder = createFolderIfNecessary(
+                HASD_PARTICIPANTS_FOLDER_NAME + EXTRACTED_SUFFIX);
+        if (hasdParticipantsFolder != null && hasdParticipantsExtractedFolder != null) {
```
what does it mean for hasdParticipantsFolder or hasdParticipantsExtractedFolder to be null?  When this happens is it correct to skip the extraction (as the code does today) or should an exception be raised?  This question applies to the other functions in this class too. ... I think this if() (and all like it  in this class) should be removed, instread createFolderIfNecessary() should throw an exception if it fails.

dephillipsmichael replied:
I had similar thoughts while writing the code that I wasn't properly throwing exceptions.  I'm so used to writing crash-proof java code for Android, that this is a bit of a paradigm shift for me. I'll switch everything to throw exceptions, I see many places where this can be improved from your PR comments.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+        File folder = new File(folderName);
+        if (!folder.exists() || !folder.isDirectory()) {
+            if (!folder.mkdir()) {
+                System.out.println("Could not create folder " + folderName);
```
shouldn't this raise an exception instead of returning null?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+            @NonNull File testSessionExtractedFolder) throws IOException {
+
+        Map<String, CompletedTestList> completedTestMap = new HashMap<>();
+        Map<String, List<TestSession>> map = new HashMap<>();
```
maybe call this 'testSessionMap' instead of 'map'?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+        Map<String, List<TestSession>> map = new HashMap<>();
+
+        int migrationFileCount = 0;
+        List<String> failedToParseList = new ArrayList<>();
```
```new ArrayList<String>();```

dephillipsmichael replied:
This was a surprising comment, is this something that eclipse enforces? Android Studio complains when I add in explicit String type in cases like this.  

I think with Java 7 or later, the compiler can do generic type inference with cases like this.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+            File file = new File(filePath);
+            migrationFileCount++;
+            ObjectMapper mapper = new ObjectMapper();
+            InputStream is = new FileInputStream(file);
```
try/finally or try-with-resource, i.e. ensure every opened input stream is closed regardless of whether an exception is thrown

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+        System.out.println("Test sessions parsing complete\nParsed " + migrationFileCount + " files.");
+
+        if (!failedToParseList.isEmpty()) {
+            System.out.println("Failed to parse file(s) " +
```
Is it OK to log the failures, or should an exception be thrown?  Do the failures require a corrective action?

dephillipsmichael replied:
I think because this deals with user data, any exception should warrant immediate program exit so that no data is corrupted.

dephillipsmichael replied:
@brucehoff what type of exception would you suggest in this case?  IllegalStateException?  I'm not well versed on exception types.

brucehoff replied:
Arguably the error is an `IllegalArgumentException`, but I would not complain if you used `RuntimeException`, which is the parent class of both `IllegalArgumentException` and `IllegalStateException`.  The important thing is to retain the informative message which is currently in the print statement.  When that message is captured and sent out as a notification the recipient will know how to act because they will be told what files are invalid.

dephillipsmichael replied:
Will switch to `IllegalArgumentException`

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                    sessions = new ArrayList<>();
+                }
+                sessions.add(sessionObj);
+                map.put(sessionObj.participant_id, sessions);
```
This should go inside the if(), above.  I.e., if you just got a non-null 'sessions' you don't need to put it.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+        UserArcIdAttributes attributes = new UserArcIdAttributes();
+        attributes.ARC_ID = arcID;
+        user.attributes = attributes;
+        user.externalId = arcID;
```
Not necessary, the server will refer to your map of studyIds to externalIds.

dephillipsmichael replied:
Good to know, removed the line.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                // Iterate through the list of folders of days unzipped
+                for (String filename : filePathList) {
+                    // Ignore everything but JSON files
+                    if (filename.endsWith(".json")) {
```
do you not also have to check that `filename` isn't a directory?

dephillipsmichael replied:
Any suggestions on how to check if a String is a directory?  I thought it was a valid assumption that no folder would ever end it's name in ".json"

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/UserMigration.java:
```diff
@@ -0,0 +1,73 @@
+            // TODO: mdephillips 9/22/21 create users on bridge
+
+        } catch (IOException | SynapseException e) {
+            System.out.println("Failed to migrate data " + e.getLocalizedMessage());
```
Don't catch the exceptions, just let them fly.  You can put `SynapseUtil.clearAllFiles();` in a `finally {}` block and/or call it at the **before** everything else (to make sure you start clean).

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/DataMigration.java:
```diff
@@ -0,0 +1,88 @@
+            System.out.println("Completed data migration\nMigrated " + uniqueUserData.size() + " accounts.");
+
+        } catch (IOException | SynapseException e) {
+            System.out.println("Failed to migrate data " + e.getLocalizedMessage());
```
Don't catch the exceptions, just let them fly.  You can put `SynapseUtil.clearAllFiles();` in a `finally {}` block and/or call it at the **before** everything else (to make sure you start clean).

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/DataMigration.java:
```diff
@@ -0,0 +1,88 @@
+ * in HM's JSON export format, parses the JSON,
+ * and then uploads the data to the corresponding Bridge users.
+ *
+ * This must be run AFTER the UserMigration executable to succeed.
```
Does this repo' have a ReadMe that explains this along with any other instructions for running the code?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/DataMigration.java:
```diff
@@ -0,0 +1,88 @@
+        try {
+            SynapseUtil.initializeSynapse();
+
+            SynapseUtil.DianFiles files = SynapseUtil.findRelevantFiles();
```
Note to self:  Verify that findRelevantFiles() only looks at the incremental data transfer and doesn't re-process the same files each time it runs.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+    // However, to get the app to run, you must provide your own email/password,
+    // NEVER commit your credentials to github
+    public static String bridgeAdminEmail = "";
+    public static String bridgeAdminPassword = "";
```
Pass the password as an environment variable.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+    public static String bridgeAdminPassword = "";
+
+    // TODO: mdephillips 9/19/21 these will need switched to the official study created on bridge
+    private static String STUDY_ID = "dian-validation-study";
```
Pass configuration parameters as environment variables.  This will allow us to switch from UAT to production.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+    private static String STUDY_ID = "dian-validation-study";
+    private static String BRIDGE_ID = "dian-validation";
+    // TODO: mdephillips 9/19/21 this will need switched to an empty array list for production
+    private static List<String> userDataGroups = Collections.singletonList("test_user");
```
Again, any configuration info should be passed as an env' variable.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+
+    private static String bridgeApi(String url, String sessionToken, String method, String body) throws MalformedURLException, IOException {
+        // Open a connection(?) on the URL(??) and cast the response(???)
+        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
```
connection.close() should be in a `finally{}` block at the end.  Accumulating open connections is a known problem for HTTP communication.

brucehoff replied:
This issue still needs to be addressed.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/BridgeUtil.java:
```diff
@@ -0,0 +1,570 @@
+        connection.setDoOutput(true);
+
+        if (body != null) {
+            try (OutputStream os = connection.getOutputStream()) {
```
yea, try-with-resource!! 👍

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/ArcValidationUserCreator.java:
```diff
@@ -0,0 +1,102 @@
+            System.out.println("Accounts ready to create on bridge " + hmUsers.size());
+
+        } catch (IOException | SynapseException e) {
+            System.out.println("Failed to migrate data " + e.getLocalizedMessage());
```
Don't swallow the exceptions, just let them fly.  If there's a problem in production we want the whole pipeline to fail and notify someone to investigate.  Whatever framework we use to run the code will detect failures and send notifications.  This code just has to ensure the exception is raised.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/ArcValidationUserCreator.java:
```diff
@@ -0,0 +1,102 @@
+            // which mdephillips created to automate QA tester creation
+            FileHandleAssociation qaFile = new FileHandleAssociation();
+            qaFile.setAssociateObjectId("syn26235874");
+            qaFile.setFileHandleId("81987693");
```
The syn ID should be a configuration parameter (another env' variable).  The file handle ID can be retrieved from Synapse, given the syn ID, and should neither be hardcoded nor passed as a param'.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/ArcValidationUserCreator.java:
```diff
@@ -0,0 +1,102 @@
+        };
+        for (File folder: folders) {
+            if (folder.exists()) {
+                FileUtils.recursivelyDeleteDirectory(folder);
```
Why not run this at the beginning to ensure that you are starting clean?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+            File file = new File(filePath);
+            migrationFileCount++;
+            ObjectMapper mapper = new ObjectMapper();
+            InputStream is = new FileInputStream(file);
```
try-with-resource

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                    // There are multiple schedule entries per user,
+                    // So we want to get the most recent one by looking at the filename
+                    // which will always end with an iso 8601 date and ".json"
+                    String existingFilePath = existingFile.getAbsolutePath();
```
I would like to see this code factored out into a separate function and tested with a variety of `<existingFileName, newFileName>` string pairs.  I don't totally understand it and the test cases should make it clear.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+        System.out.println("Test sessions parsing complete\nParsed " + migrationFileCount + " files.");
+
+        if (!failedToParseList.isEmpty()) {
+            System.out.println("Failed to parse file(s) " +
```
What do the failures mean?  Do they indicate that manual intervention is needed?  Should we throw an exception and stop the pipeline?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+
+        try {
+            // Find folders
+            EntityChildrenResponse response = synapse.getEntityChildren(request);
```
This returns the first of potentially multiple pages.  You have to use `response.getNextPageToken()` to see if there are more pages.

dephillipsmichael replied:
Is there a function in the SDK to get all pages?  Or, is there a way to set the max page size to something like 365?  In this case that would be a year's worth of daily folders, and we would never hit the max, as this code should only live for a max of 3 months.

dephillipsmichael replied:
I actually wrote this function that seems to do the job:

```
  /**
     * @param parentId of the folder to look in for EntityHeaders
     * @param type of files to look for, usually folder or file
     * @return the full list of entity headers on all pages of result
     */
    public static @NonNull List<EntityHeader> getAllEntityChildren(
            String parentId, EntityType type) throws SynapseException {

        List<EntityHeader> entityHeaderList = new ArrayList<>();

        EntityChildrenRequest fileRequest = new EntityChildrenRequest();
        fileRequest.setParentId(parentId);
        fileRequest.setIncludeTypes(Lists.newArrayList(type));

        while (fileRequest != null) {
            EntityChildrenResponse fileResponse = synapse.getEntityChildren(fileRequest);
            if (fileResponse.getPage() != null) {
                entityHeaderList.addAll(fileResponse.getPage());
            }
            String nextPage = fileResponse.getNextPageToken();
            if (nextPage == null) {
                fileRequest = null;
            } else {
                fileRequest.setNextPageToken(nextPage);
            }
        }

        return entityHeaderList;
    }
```

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+    /**
+     * @return all the relevant ZIP files to download for the migration
+     */
+    public static DianFiles findRelevantFiles() {
```
How do you ensure that you don't reprocess the same files, over and over, each time you run this code?  We only want to process the incremental data transfer from HM.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                EntityChildrenResponse subFolderResponse = synapse.getEntityChildren(subFolderRequest);
+
+                // Each folder at this point, should have 3 child folders,
+                // test_session, test_session_schedule, and wake_sleep_schedule
```
OK, then check that subFolderResponse has no `nextPageToken` and raise an exception if it does.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                for (String zipFolderName: IMPORTANT_FOLDERS) {
+                    EntityHeader zipHeader = findFolderWithName(
+                            zipFolderName, subFolderResponse);
+                    if (zipHeader != null) {
```
if zipHeader is null shouldn't we raise an exception?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                                wakeSleepSchedules.add(zipFile);
+                            }
+                        } else {
+                            System.out.println("Could not find a zip file in folder" + zipFolderName);
```
raise an exception?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                            System.out.println("Could not find a zip file in folder" + zipFolderName);
+                        }
+                    } else {
+                        System.out.println("Could not find expected folder " +
```
raise an exception?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+            files.wakeSleepSchedules = wakeSleepSchedules;
+
+        } catch (Exception e) {
+            System.out.println("Failed to find all the ZIP files for testSessions, " +
```
Wrap the exception and raise it.e.g., 
```
throw new RuntimeException("Failed to find all the ZIP files ...", e);
```

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+            fileRequest.setParentId(project.getId());
+            fileRequest.setIncludeTypes(Lists.newArrayList(EntityType.file));
+
+            EntityChildrenResponse fileResponse = synapse.getEntityChildren(fileRequest);
```
This just returns the first page.  Use `fileResponse.getNextPageToken()` to get the next page.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+    /**
+     * @return all the relevant participant ZIP files to download for the migration
+     */
+    public static DianParticipantFiles findRelevantParticipantFiles() {
```
How do you ensure you just process the latest data upload and not the entire history of transferred files?

dephillipsmichael replied:
Currently, the algorithm works by processing the entire history of the HM data each time the code runs.  We did briefly discuss only processing "new" data; however, I found the algorithm was simplified by having it run the same way each time, so I wouldn't have to handle saving/loading the algorithm's state.  

Happy Medium will still only push each day's data into separate folders, so there shouldn't be any extra data accumulating in Synapse.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+            }
+
+        } catch (Exception e) {
+            System.out.println("Failed to find all the ZIP files for the participant JSON. " +
```
wrap the exception and raise it

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                }
+                userList.add(user);
+            } else {
+                System.out.println("Failed to create hm user with table id " + userAndSiteLoc.participant);
```
Raise exception

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+            return null;
+        }
+        for (File file: fileList) {
+            if (file.getName().contains(partOfName)) {
```
also check that file is not a directory?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+        EntityChildrenRequest testSessionZipRequest = new EntityChildrenRequest();
+        testSessionZipRequest.setParentId(baseFolder.getId());
+        testSessionZipRequest.setIncludeTypes(Lists.newArrayList(EntityType.file));
+        EntityChildrenResponse testSessionZipResponse = synapse.getEntityChildren(testSessionZipRequest);
```
This only gets the first page.  If there is no zip file on the first page and there is a zip file in a later page then this will erroneously conclude there is no zip file.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+        }
+
+        for (EntityHeader folder: response.getPage()) {
+            if (name.equals(folder.getName())) {
```
This does not check that the entity is a folder.  It would return a non-folder entity in `response` if it has the right name.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+                if (testSession.finished_session != 1) {
+                    unique = false; // skip unfinished schedules
+                }
+                for (CompletedTest completedTest: completed) {
```
seems like this `for` loop should be contained in an `else`, i.e. if not finished don't bother comparing to `completedTest`.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,970 @@
+
+        public String participant_id;
+        public String os_type;
+        public int finished_session; // 1 == finished, 0 == unfinished
```
why not use a boolean here instead of an int?

dephillipsmichael replied:
This is Happy Medium's data model, I can't change it at this point.

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+    private static SynapseClient synapse;
+
+    private static Project project;
+    private static String projectId = "syn25791302";
```
I'd rather see the test set up its own test data than assume data is set up in some Synapse project.

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        // Create 2 users
```
What does "Create 2 users" mean?

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+    public static void beforeClass() throws Exception {
+        // Create 2 users
+        synapse = new SynapseClientImpl();
+        synapse.setBearerAuthorizationToken(SynapseUtil.synapsePersonalAccessToken);
```
The test could get an access token as an env' var.

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+        EntityChildrenResponse response = synapse.getEntityChildren(request);
+        assertNotNull(response);
+        assertNotNull(response.getPage());
+        assertEquals(2, response.getPage().size());
```
Not clear what `testNavigateFolderStructure` is testing.  There's no reference to the logic in SynapseUtil and this code need not test that the Synapse client works.

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+    }
+
+    @Test
+    public void testDownload_testSession_ZIP() throws SynapseException, IOException {
```
Not clear what testDownload_testSession_ZIP is testing. There's no reference to the logic in SynapseUtil and this code need not test that the Synapse client works.

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+    }
+
+    @Test
+    public void testDownload_SessionSchedule_Zip() throws SynapseException, IOException {
```
Not clear what testDownload_SessionSchedule_Zip is testing. There's no reference to the logic in SynapseUtil and this code need not test that the Synapse client works.

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+    }
+
+    @Test
+    public void testDownload_WakeSleep_Zip() throws SynapseException, IOException {
```
Not clear what testDownload_WakeSleep_Zip is testing. There's no reference to the logic in SynapseUtil and this code need not test that the Synapse client works.

brucehoff commented on DataMigration/src/test/java/org/sagebionetoworks/dian/datamigration/SynapseTests.java:
```diff
@@ -0,0 +1,307 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.sagebionetoworks.dian.datamigration;
```
I see very little test coverage for SynapseUtil, just two methods?  Much more unit testing is needed.

dephillipsmichael committed:
Major refactor into smaller classes with improved unit test coverage

dephillipsmichael committed:
Update for docker test

dephillipsmichael committed:
Push to docker test github action

dephillipsmichael committed:
Fixed broken test

dephillipsmichael committed:
Debugging help on failing CI unit test

dephillipsmichael committed:
More unit test failure debugging

dephillipsmichael committed:
Fixed ZIP unit test by adding missing file

dephillipsmichael committed:
Working docker file

dephillipsmichael committed:
Added new UserMigration module

dephillipsmichael committed:
Fixed misspelled package name from bionetoworks to bionetworks

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/DataMigration.java:
```diff
@@ -0,0 +1,138 @@
+    private static void runUserMigration() throws SynapseException, IOException {
+        System.out.println("Beginning User Migration");
+
+        // TODO: mdephillips 10/2/21 testing docker build, uncomment after
```
Please fix the commented out code.

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/DataMigration.java:
```diff
@@ -0,0 +1,138 @@
+    private static void runDataMigration() throws SynapseException, IOException {
+        System.out.println("Beginning Data Migration");
+
+        // TODO: mdephillips 10/2/21 testing docker build, uncomment after
```
fix commented out code

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/FileHelper.java:
```diff
@@ -0,0 +1,135 @@
+    /**
+     * @param root containing the file to find
+     * @param partOfName to look for in the filename for a match
+     * @return the file if there was a match, null otherwise
```
"the file" -> "the first file or directory found that matches 'partOfName'"

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/FileHelper.java:
```diff
@@ -0,0 +1,135 @@
+
+    /**
+     * @param extractedFolder folder containing a list of folders, that contain JSON files
+     * @return a list of the file paths to all the JSON files in the extractedFolder
```
"all the JSON files in the extractedFolder" -> "all the JSON files in some folder within the extractedFolder"
(I.e. any JSON files that are immediate children of 'extractedFolder' are ignored, correct?)

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/FileHelper.java:
```diff
@@ -0,0 +1,135 @@
+                String[] expectedDirPathList = file.list();
+                if (expectedDirPathList != null) {
+                    filePathList = expectedDirPathList;
+                }
```
You could make this code much more terse by omitting `filePathList` and just having:
```
if (expectedDirPathList==null) {
   continue;
}

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/HmDataModel.java:
```diff
@@ -0,0 +1,293 @@
+    public static class TableRow {
+
+        public static SiteLocation findSiteLocation(
+                String id, ParticipantSiteLocation[] siteLocMap, SiteLocation[] sites) {
```
`id` -> `participantId`

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,447 @@
+                .findAllJsonFilesInFolder(testSessionScheduleExtractedFolder);
+
+        // The ZIPs have json files that all follow the format at the end of the name like this
+        int filenameDateSize = "2020-02-20T12-31-13Z.json".length();
```
`filenameDateSize ` should be a static variable

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,447 @@
+     * @return the participant ID with a length of 6, with leading "0"s added
+     */
+    public static String fixParticipantId(String participantId) {
+        int targetLength = 6;
```
`targetLength` should be a static variable

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,447 @@
+    public static String fixParticipantId(String participantId) {
+        int targetLength = 6;
+        if (participantId.length() >= targetLength) {
+            return participantId.substring(0, targetLength);
```
instead, `throw new IllegalArgumentException("Expected string length <=6 but found "+participantId);`

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,447 @@
+            return participantId.substring(0, targetLength);
+        }
+        int zerosToAdd = targetLength - participantId.length();
+        String zerosPrefix = StringUtils.repeat("0", targetLength).substring(0, zerosToAdd);
```
why not just `String zerosPrefix = StringUtils.repeat("0", zerosToAdd);` ?

brucehoff commented on DataMigration/src/main/java/org/sagebionetoworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,447 @@
+    public static String getRandomNumberString() {
+        // It will generate 6 digit random Number.
+        // from 0 to 999999
+        Random rnd = new Random();
```
`rnd` should be a static class variable, not a variable in `getRandomNumberString()`, initialized each time the method is called.

brucehoff commented on README.md:
```diff
@@ -1,2 +1,135 @@
+Before you can run the code, your environment must have the following environmental variables set first:
+
+## Required
+**BR_EMAIL** - Your email to sign into the Arc Validation or DIAN bridge project
```
Formatting: add line break after each parameter.  (Currently no line breaks appear when the markdown is rendered. To see the markdown appearance in the Github PR, click 'View file', at right.)

dephillipsmichael committed:
DataMigration re-factor for final draft, increased unit test coverage.

dephillipsmichael committed:
Updated ReadMe.md

dephillipsmichael committed:
Adjust docker CI

dephillipsmichael committed:
Adjusted docker build action

dephillipsmichael committed:
Updated readme to reflect docker build changes

dephillipsmichael commented:
@brucehoff the migration code PR here is ready for review.  One major change is that I switched over to Sage's Bridge Java SDK instead of using my own custom code.  This might need another quick review by @alxdarksage as well.

dephillipsmichael committed:
Fixed impossible unit test, empty folders cannot be committed to github

dephillipsmichael committed:
Removed temporary fix

dephillipsmichael committed:
Fix participant ID during participant parsing

dephillipsmichael committed:
Switched to a 9 character password

dephillipsmichael committed:
Update readme

dephillipsmichael committed:
Update SIGN_IN_TOKEN to VERFICIATION_CODE

brucehoff commented on README.md:
```diff
@@ -1,2 +1,171 @@
+**Arc ID:** This is a 6 digit unique identifier used to keep track of a participant.  In Sage Bridge terms, this is a user’s External ID.
+
+**Rater ID:** A rater is an individual who is responsible and accountable for setting up participants with the app.  A rater is important to the audit trail process so it is known who is responsible for each user.  When HM creates a rater on their dashboard, the server would generate a 6 digit rater id and then email that as plain text to the rater’s email.  This Rater ID is used as the authentic ...
+
```
I don't know what "Map Arc" or "DIAN OBS" are.  Are they mobile app's?

dephillipsmichael replied:
Map Arc is a synonym for HASD.  WashU team and the app stores call it Map Arc, but all the hM developers call it HASD, along with all the code references.  Same goes with EXR == DIAN OBS

brucehoff commented on README.md:
```diff
@@ -1,2 +1,171 @@
+These users will be created the same way as the section above, if the user has no Device ID.
+
+# Account parsing from HM's data export
+HM exports all the participant data needed to do this process.  The staging QA data is available on Synapse Synapse.  The two relevant files are:
```
repeated word 'Synapse'

brucehoff commented on README.md:
````diff
@@ -1,2 +1,171 @@
+sage_qa_staging_participant_rater-9-21-21.json
+```
+
+The migration code will download and unzips these files from Synapse, parse the JSON, and build the data models representing each individual user that exists on HM's servers.
````
unzips -> unzip

brucehoff commented on README.md:
```diff
@@ -1,2 +1,171 @@
+As outlined earlier, HM communicated Arc ID and Rater ID passwords through email.  We are no longer supporting that.  Each site location manager will need to create a Synapse account.  After they create a Synapse account, they will be given access to the DIAN bridge project, and only given permission to view their sub-study and its participants.  WashU will not be able to view any of the sub-stud ...
+
+# Account Migration to Sage's Bridge Server
+Before the migration can complete successfully, a sub-study must be created for each site location on Bridge.  Each user we create, must belong to a sub-study.
```
extra comma: "Each user we create must belong to..."

brucehoff commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+# HM Data Model
+### Wake Sleep Schedule
+The availability schedule defines when a user is available to do their testing. It is a JSON model, and while there can be multiple per user.  We only need to read the most recent.
```
"...per user, we only need to ..."

brucehoff commented on README.md:
```diff
@@ -1,2 +1,171 @@
+The availability schedule defines when a user is available to do their testing. It is a JSON model, and while there can be multiple per user.  We only need to read the most recent.
+
+### Test Schedule
+The test session schedule defines when a user is available to do their testing. It is a JSON model, and while there can be multiple per user.  We only need to read the most recent.
```
This repeats the definition of 'Wake Sleep Schedule' (including the typo).

dephillipsmichael replied:
Good catch, fixed.

brucehoff commented on README.md:
```diff
@@ -1,2 +1,171 @@
+**SYN_PAT** - Your Synapse access token to your account that has read/download access for the dian uat or dian prod Synapse projects.
+**SYN_PROJ_ID** - The Synapse Project ID for dian uat or dian prod.
+
+# Building the code
```
Maybe change this to 'Building and running the code' and say how to run in each form.  For example, in docker is it as simple as `docker run -e BR_PW:<pw> -e .... ghcr.io/sage-bionetworks/diandatamigration:release-test1`? Or do you also have to specify an entry point, args or something? 

Also, _when_ should the code be run and what preconditions need to be set up? Do we have to coordinate with HM's dump?  Does the BRIDGE project need to be set up in some way before running the code?   Is the code idempotent (if we run it twice will it put the Bridge project into a broken state)?

dephillipsmichael replied:
This is definitely valuable info that I should add.  I'm releasing a new package on branch release/test2, and I'll update the ReadMe after I get the docker info I need.  There should be no args or entry point besides the environmental variables.

The _when_ is a good question.  I think that is something we will need to coordinate with HM.  I'll sync with Ann on this and get back to you on adding this to the ReadMe. For now, I'll say an hour after HM's data dump. 

Also, yes, the code is idempotent, even if you ran it twice at the same time, it would be fine, just inefficient.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+     * @param data app data associated with the user, from HappyMedium's servers
+     * @throws IOException if something goes wrong make server calls to Bridge
+     */
+    public static void migrateUser(HmDataModel.HmUser user,
```
I don't see any test for this. Ideally there would be a unit test using mocked API objects.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+
+    private static UserSessionInfo userSessionInfo;
+    private static ClientManager clientManager;
+    private static ForResearchersApi researcherApi;
```
These should be mockable, for testing.  In Synapse we do so using Spring's dependency injection.  If you do not want to introduce an additional framework to your code base you can achieve this by (1) making them NOT static, (2) adding a constructor that allows passing them in, (3) adding a default constructor that does what `initialize()` does today.  You would then call the constructor mentioned in (2) in the 'setup' of your unit test, passing mocked versions of the objects created using Mockito.

alxdarksage replied:
Mockito has really nice annotations for this, see for example https://github.com/Sage-Bionetworks/BridgeServer2/blob/develop/src/test/java/org/sagebionetworks/bridge/services/OrganizationServiceTest.java

dephillipsmichael replied:
Oh nice! I see that you've given me the guidance I needed already, so ignore the other comments about testing the functions in this class.  I'll look at mocking these and making all the functions in this class testable.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+     * @param data associated with the user, migrated from HappyMedium's server
+     * @throws IOException if something goes wrong creating the account, or writing reports
+     */
+    private static void createMigrationAccountWithData(
```
I don't see a test for this.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+    private static void updateMigrationAccount(
+            StudyParticipant participant, HmDataModel.HmUser user,
+            HmDataModel.HmUserData data) throws IOException {
+        System.out.println("Updating migration data for user " + user.externalId);
```
The `user` parameter isn't used except in the print statement.  Is that correct?  Is this covered by a test?

dephillipsmichael replied:
Same response as response to writeUserReports comment.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+     * @param user HappyMedium user
+     * @throws IOException if something goes wrong deleting user report data and attributes
+     */
+    private static void clearMigrationData(
```
The `user` parameter isn't used except in the print statement. Is that correct? Is this covered by a test?

dephillipsmichael replied:
The user being passed only for the print statement is intentional. 

It is not covered by a test, but I have the same response as I did with your writeUserReports comment.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+     * @param data that will be uploaded as user reports
+     * @throws IOException if something goes wrong writing the reports to Bridge.
+     */
+    private static void writeUserReports(String userId, HmDataModel.HmUserData data) throws IOException {
```
Is this covered by a test?

dephillipsmichael replied:
It is not currently, and I'm not experienced with unit testing functionality like this. 

Would it be testing that the different calls to `reportsApi.addParticipantReportRecordV4` are all being invoked when they are supposed to?  Do you have an example of how to do something like that?

brucehoff replied:
"Do you have an example of how to do something like that?"  Yes.  First of all you would not test this logic along with the logic of addParticipantReportRecordV4(), makeReportData(), and writeValueAsString() all in the same test.  The other methods should be tested separately and here we would simply assume they work correctly. Here you would *mock* reportsApi and simply verify that addParticipantReportRecordV4() is called with the correct inputs.  Ditto for objectMapper.  You would check that objectMapper.writeValueAsString() is passed the correct input.

dephillipsmichael replied:
Appreciate the feedback, looks like it won't be too difficult to make this happen.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/HmDataModel.java:
```diff
@@ -0,0 +1,411 @@
+ * as well as the data models we use to transfer HM users
+ * over to Bridge server.
+ */
+public class HmDataModel {
```
Need tests for the six TableRow.find*() methods and the CompletedTestList() constructor.

dephillipsmichael replied:
Another good catch, I appreciate the thoughtful review.  I added tests for all of those functions, they will be in a new file called `HmDataModelTests`

brucehoff commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/MigrationTests.java:
```diff
@@ -0,0 +1,386 @@
+        assertEquals("000123", MigrationUtil.fixParticipantId("123"));
+        assertEquals("001234", MigrationUtil.fixParticipantId("1234"));
+        assertEquals("012345", MigrationUtil.fixParticipantId("12345"));
+        assertEquals("123456", MigrationUtil.fixParticipantId("123456"));
```
test the truncating case, "1234567".

dephillipsmichael replied:
Ahhh of course, good catch. Done.

brucehoff commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/MigrationTests.java:
```diff
@@ -0,0 +1,386 @@
+        participantPathList.add(participantsFolder);
+
+        List<HmUser> users = MigrationUtil
+                .createHmUserRaterData(participantPathList);
```
Does this test the case that a path in the list is null?

dephillipsmichael replied:
Good thought... I didn't test that in particular, but I added a case to this test to check that it properly throws a NullPointerException when there is a null path in the list, which is the desired behavior.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,438 @@
+            ParticipantSiteLocation[] participantSiteLocList = TableRow.parseTableRow(mapper, pathList.get(
+                    ParticipantFileEnum.PARTICIPANT_SITE_LOCATION), ParticipantSiteLocation[].class);
+
+            for (Participant user : participantList) {
```
There's a potential NullPointerException, i.e., if participantList is null. If you don't want to wrap the `for` loop in another `if`, then change `TableRow.parseTableRow` to return `Optional`, i.e., `participantList` would have type `Optional<Participant[]>`.  Then change this line to:
```
for (Participant user : participantList.orElse(new Participant[0]))
```

dephillipsmichael replied:
Usually I would agree here and do a null check.  However, if there are no participants created, there is something wrong either with the export from HappyMedium or my migration code, and I would prefer this to crash. Let me know if you feel strongly otherwise.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,438 @@
+        return data;
+    }
+
+    public static void assignStudyId(HmUser user, SiteLocation site,
```
I don't see a test for this.

dephillipsmichael replied:
I moved it into the HmUser constructor, and made a few unit tests for the constructor.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,438 @@
+                userMatch.arcId = fixParticipantId(participant.participant_id);
+
+                // Assign the user's Study ID based on various parameters
+                assignStudyId(userMatch, site, deviceId, note);
```
This is kind of an odd pattern.  Why not create a constructor or factory that returns a new HmUser, and then fill in other fields, e.g.:
```
String arcId = fixParticipantId(participant.participant_id);
HmUser userMatch = createHmUser(arcId, site, deviceId, note);
userMatch.rater = rater;
...
```

dephillipsmichael replied:
Totally agree, will switch over to a constructor model.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/MigrationUtil.java:
```diff
@@ -0,0 +1,438 @@
+     * @param siteName to convert to a bridge acceptable site name
+     * @return the converted site name
+     */
+    public static String bridgifySiteName(String siteName) {
```
Add a test.

dephillipsmichael replied:
Done

brucehoff commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/PathsHelperTests.java:
```diff
@@ -0,0 +1,176 @@
+        }
+        assertNotNull(e);
+
+        PathsHelper.deleteDirectoryRecursively(folderUnitTestsDir);
```
If this is done in the after(), it isn't needed here.

dephillipsmichael replied:
Good catch, removed and verified you are correct.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/PathsHelper.java:
```diff
@@ -0,0 +1,184 @@
+     * @return a list of files within the directory
+     * @throws IOException if something goes wrong accessing the directory's files
+     */
+    public static List<Path> getFilesInDirectory(Path directory) throws IOException {
```
Does this not also return sub-directories? Is that what's desired?

dephillipsmichael replied:
It does not return sub-directories, and is not desired.  See test `PathsHelperTests.test_getFilesInDirectory()`

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/SecureTokenGenerator.java:
```diff
@@ -0,0 +1,137 @@
+
+        for (int i = 0; i < password.length(); i++) {
+            String character = Character.toString(password.charAt(i));
+            containsUppercase |= UPPERCASE_ALPHA.contains(character);
```
I believe `|=` is the bitwise operator while `||=` is the boolean operator.

dephillipsmichael replied:
This has been replaced by Alx's PasswordGenerator.

However, it is a good catch. I think this worked because Java booleans are stored as just bit 1 or bit 0, but you are right that it is the bitwise OR and not logical OR.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/SecureTokenGenerator.java:
```diff
@@ -0,0 +1,137 @@
+     * Guaranteed to generate a random password string compatible with Bridge
+     * @return a random password string compatible with Bridge
+     */
+    public String nextBridgePassword() {
```
Generating passwords randomly until you get a valid one is only OK if it only takes a few iterations.  As a sanity check I computed the probability of generating a password with no digits:  There are 74 characters, so 65 of them are not digits. The probability of selecting a non-digit character is 65/74 and the probability of doing that nine times is (65/74)^9 = 0.31.  Repeating four times generates a digit-containing password with 99% probability.  So while it may take a few retries, it's probably not going to be more than 10.  You might put an upper limit, e.g., after 100 tries, throw a RuntimeException().

alxdarksage replied:
See https://github.com/Sage-Bionetworks/BridgeServer2/blob/b4a0fb6bfcd52577dd72d744ca8dfa72f84b9631/src/main/java/org/sagebionetworks/bridge/PasswordGenerator.java

We generate a random password, then randomly insert one of every kind of special character somewhere in the string to ensure it passes any password validation set for a specific app, allowing us to generate a password in one pass.

dephillipsmichael replied:
Nice! Thank you. 

First, I love that you did those calculations @brucehoff a true computer scientist!

Also, I switched over to Alx's PasswordGenerator class.  It is an order of magnitude more efficient than mine, especially when run at 1,000,000 iterations.  Doing a second pass using random indices of the password is a smart way to guarantee it meets the criteria.

One thing I did notice however was that the PasswordGenerator's passwords only ever have exactly one SYMBOLIC character. I wanted to double check that should be fine for DIAN? @alxdarksage  I could switch it to include the SYMBOLIC characters on the first pass; however, that could lead to upwards of 6 symbolic characters in my 9 character password, and that seems like bad user experience to type in.

alxdarksage replied:
I think our generator uses all symbols in the first pass and then ensures there's at least one, but because you have users typing these in, it is probably fine to only have one of the special characters. Again, this is still vastly more secure than what we were going to do initially!

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    /**
+     * @param parentId of the folder to look in for EntityHeaders
```
parentId ID of the folder to look in for EntityHeaders

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,287 @@
+            }
+            String nextPage = fileResponse.getNextPageToken();
+            if (nextPage == null) {
+                fileRequest = null;
```
This is just personal preference but I would change this line from `fileRequest=null;` to `break;`.  The start of the loop is then just `while(true) {`

dephillipsmichael replied:
I'm not used to utilizing the java language feature "break;", but this seems like the correct way to do this.

brucehoff commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/SynapseUtil.java:
```diff
@@ -0,0 +1,287 @@
+        return null;
+    }
+
+    public static class DianFiles {
```
is this used anywhere?  I don't see a reference.

dephillipsmichael replied:
Good catch, deleted.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+
+        AuthenticationApi authApi = clientManager.getClient(AuthenticationApi.class);
+        UserSessionInfo sessionInfo = authApi.signInV4(signIn).execute().body();
+        userSessionInfo = clientManager.getSessionOfClients();
```
I believe sessionInfo and userSessionInfo should be the same.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+
+        // Set the new attributes to be blank, except for IS_MIGRATED = true
+        StudyParticipant newParticipant = new StudyParticipant()
+                .attributes(MIGRATED_USER_ATTRIBUTES);
```
You should be able to set a map with the one value you want to retain, rather than setting fields to blank strings. If you can't, I would consider that a bug on the server. The lack of a value in the JSON should always be interpreted as a delete/nullification of the value.

dephillipsmichael replied:
Good to know!  Will update.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+        String notes = "";
+        if (user.notes != null) {
+            notes = user.notes;
+        }
```
Really minor but why not just add keys to the map if they exist, rather than setting them to blank strings? The presence of fields intended to indicate "no value" may cause you headaches down the road.

dephillipsmichael replied:
Ahhh thank you for this comment.  I didn't know that the lack of a field translates to "no value" - I'll switch over to that.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+        reportData.setLocalDate(REPORT_DATE);
+        reportData.setData(clientData);
+        return reportData;
+    }
```
As mentioned the participant data API is a similar JSON store, but not for time series data. You can just save one item under an identifier without the date/time headaches.

dephillipsmichael replied:
I was very tempted to switch over to that, but at this point it would be too much work and too risky to update the Android/iOS apps to use that instead; however, this is definitely going to be my approach for future projects.

alxdarksage commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/HmDataModel.java:
```diff
@@ -0,0 +1,411 @@
+                    String.valueOf(day) +
+                    String.valueOf(session)).hashCode();
+        }
+    }
```
There is a library called EqualsVerifier that will verify your equality and hash code methods are robust, if the value identity of these objects is critical to your program logic. For example, the fact that fields are not final can cause issues.

dephillipsmichael replied:
Ohhh nice suggestion, I'll look into testing this.

dephillipsmichael replied:
I brought the library in, but I noticed it it wants week, day, and session to be final, which doesn't seem to play nicely with how jackson json library does decoding.  I'm going to switch to manually testing equals/hashCode.

alxdarksage commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/PathsHelperTests.java:
```diff
@@ -0,0 +1,176 @@
+    @After
+    public void after() throws IOException {
+        PathsHelper.deleteDirectoryRecursively(folderUnitTestsDir);
+    }
```
Is this unit test writing to a directory? I feel like this should be mocked and actual filesystem manipulation should be tested through an integration test.

dephillipsmichael replied:
Yes, the unit tests do make temporary filesystem changes, and I tested that even if a unit test fails or crashes, the filesystem is cleaned up in the after() func.  This has been tested both locally and on the github environment.

Java isn't my strongest language, so this may be obvious, but is your comment a conceptual/architectural issue, or is there something different about the way java unit tests are run vs. integration tests?

alxdarksage replied:
IMO unit tests should mock all external dependencies, including things like the filesystem. They should run entirely in memory, they should test one class at a time—thoroughly. Integration tests verify that the components (like your code and the operating system) work together correctly to fulfill a function.

alxdarksage commented on DataMigration/src/test/resources/participants/sage_qa_staging-participant_device-10-11-21.json:
```diff
@@ -0,0 +1,80 @@
+        "os_type": "Android",
+        "raw_device_info": "Android|Google Pixel 2|10|29|google\/walleye\/walleye:10\/QQ1A.191205.008\/5974827:user\/release-keys"
+    }
+]
```
Do you test if fields of these JSON files are null, or missing? Do you ever get empty values and are these properly converted to nulls (if that's their semantic meaning)? I am not reviewing the tests in detail so maybe there's another way you're verifying robustness in the face of bad inputs.

dephillipsmichael replied:
There are various optional fields in HappyMedium's JSON model.  One example is a user's site location contact_name and contact_email.  The jackson JSON parser assigns null to them in the case of missing fields. 

If a field is required by the migration code, and it is missing from the data, a NullPointerException will be thrown as I don't do any null conditional checks by design.

However, I will add some unit tests that make sure the NullPointerException is thrown, that way we know 100% that this will be the case.

alxdarksage replied:
Okay, if it's supposed to fail, then more work on this isn't necessary.

alxdarksage commented on .idea/.name:
```diff
@@ -0,0 +1 @@
+DataMigration
```
I wouldn't normally check in IDE project files.

dephillipsmichael replied:
Over 90% of the project files are in the .gitignore, but this is one that should actually be included in source control, for AndroidStudio at least.

alxdarksage replied:
There were a number of .idea files and I'm not sure why the .idea folder would be in source control (that's IntelliJ project files, right?)

alxdarksage commented on DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtil.java:
```diff
@@ -0,0 +1,314 @@
+    private static final String ATTRIBUTE_PHONE_NUM = "PHONE_NUMBER";
+    private static final String ATTRIBUTE_IS_MIGRATED = "IS_MIGRATED";
+    private static final String ATTRIBUTE_VALUE_FALSE = "false";
+    private static final String ATTRIBUTE_VALUE_TRUE = "true";
```
There are some assumptions about the Bridge app configuration that are made by this migration script, I think. I don't know if it's worth mentioning what they are in your documentation (and you may have already—in which case, ignore this comment).

dephillipsmichael replied:
It should be covered in the ReadMe I wrote that Bridge needs the attributes added, as well as all the sub-studies properly setup before the migration code will work.

dephillipsmichael committed:
Pull request feedback

dephillipsmichael committed:
ReadMe updates

dephillipsmichael committed:
Fixed package error

dephillipsmichael committed:
ReadMe updates

dephillipsmichael committed:
Added test coverage for BridgeJavaSdkUtil class

dephillipsmichael commented:
@brucehoff @alxdarksage I think I've addressed all the comments and made the necessary updates.  This is ready for review again.  I re-ran both the jar, and the docker image and both succeeded as expected.

Test coverage has been increased with the new class `HmDataModelTests` and major improvements for `BridgeJavaUtilSdkTests`

<img width="947" alt="Screen Shot 2021-11-01 at 7 27 40 PM" src="https://user-images.githubusercontent.com/5590748/139776817-382e9bc6-0201-4615-9511-6cf03178397f.png">

brucehoff commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtilTests.java:
```diff
@@ -0,0 +1,455 @@
+
+    @Before
+    public void before() throws IOException {
+        if (mockResearcherApi == null) {  // Only initialize once
```
If you want to set up something once per class (rather than once per test) the way to do it with JUnit is to use the `@BeforeClass` annotation.  (Note that the annotated method is *static*.)  I'm not sure why you want to set up the mocked objects just once per class though.  We normally set up all the mocks before each test.

dephillipsmichael replied:
I avoided `@BeforeClass` because it was _static_.  I'm new-ish to Mockito, so I thought it was bad practice to setup before each test, but I removed the null conditional and it works the same and does not same any less efficient.

brucehoff replied:
Setting up before each test is good.  It makes sure each test starts from a known state.

brucehoff commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/BridgeJavaSdkUtilTests.java:
```diff
@@ -0,0 +1,455 @@
+        HmDataModel.HmUserData data = createFullNewUserData(user.arcId);
+
+        BridgeJavaSdkUtil.migrateUser(user, data);
+        verify(mockSignUpCall).execute();
```
FYI: We have switched to JUnit 5.  Among its features is increased strictness.  Tests fail if their mocks are not invoked. It forces you to more precisely set up the mocks for each test but the upside is that you don't have to add 'verify()' to check that your mock is called.  You get that automatically!

dephillipsmichael replied:
This sounds awesome. It does look like I'm targeting JUnit5, maybe I'm missing something with my setup?

`testImplementation("org.junit.jupiter:junit-jupiter:5.8.0")`

Is it because I am using this Mockito version?

`testImplementation("org.mockito:mockito-core:2.23.4")`

brucehoff replied:
I know you are not using JUnit 5 because your imports look like "org.junit...", while JUnit 5 imports look like "org.junit.jupiter...". Also, JUnit 5 uses `@BeforeEach` rather than `@Before`.  It's fine for this PR. I just wanted to pass along the "FYI."

brucehoff commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/MigrationTests.java:
```diff
@@ -0,0 +1,419 @@
+        } catch (NullPointerException e) {
+            nullExceptionCaught = true;
+        }
+        assertTrue(nullExceptionCaught);
```
FYI You can assert that an exception was thrown.

dephillipsmichael replied:
Another new trick I've learned from you, thanks!  Switched to this:

```
        assertThrows(NullPointerException.class, () ->
                MigrationUtil.createHmUserRaterData(participantPathList));
```

brucehoff commented on DataMigration/src/test/java/org/sagebionetworks/dian/datamigration/PasswordGeneratorTests.java:
```diff
@@ -0,0 +1,42 @@
+
+        for (int i = 0; i < password.length(); i++) {
+            String character = Character.toString(password.charAt(i));
+            containsUppercase |= PasswordGenerator.UPPERCASE.contains(character);
```
On these four lines I believe the correct operator for boolean disjunction should be `||=`, not `|=`, which is the bitwise operator.

dephillipsmichael replied:
Oh right! Apparently ||= doesn't exist, at least in Java 8.  I switched to this format:

`containsUppercase = containsUppercase || PasswordGenerator.UPPERCASE.contains(character);`

brucehoff replied:
I was wrong.  There is no `||=`. So your change is the best sol'n.

brucehoff reviewed (changes requested):
Just a few small changes.

dephillipsmichael committed:
PR feedback

dephillipsmichael commented:
@brucehoff small changes fixed, and ready for review again
